[All Hosts] (debugging) new debugging overview article (#3112)
OfficeDev/office-js-docs-pr: docs/testing/debug-add-ins-overview.md

---
title: Debug Office Add-ins
description: 'Find the Office Add-in debugging guidance for your development environment'
ms.date: 12/02/2021
ms.localizationpriority: high
---

# Overview of debugging Office Add-ins

Debugging Office Add-ins is essentially the same as debugging any web application; but there isn't a single set of tools that all add-in developers can use. This is because add-ins can be developed on different operating systems and can run cross-platform. This article will help you find the detailed debugging guidance for your development environment.

> [!NOTE]
> This article is concerned with debugging in the narrow sense of setting breakpoints and stepping through code. For guidance on testing and troubleshooting, start with [Test Office Add-ins](test-debug-office-add-ins.md) and [Troubleshoot development errors with Office Add-ins](troubleshoot-development-errors.md).
>

> > This article uses "your development computer" and "your development environment" to refer to the environment on which you're debugging. This almost always *is* your development computer. Although you should *test* your add-in on all the platforms that you want to support, you'll only very rarely need to *debug* on an environment different from your development computer.  

Reviewer: What confusion is this note about development computer/environment trying to solve? What computer would people think we're talking about?

Author: What I'm trying to get a across with the note is that normally you don't have to debug on more than one platform. But once in a rare while you'll have to debug on an environment that is different from the one your dev machine is using. So, I want to use "your development computer" instead of "the environment on which you're debugging" because the latter gives the false impression that you have to debug on a lot of environments other than the one on your dev machine. The note is trying to explain this editorial decision. I'll rewrite.

## Server-side or client-side?

> Debugging the server-side code of an Office Add-in is exactly the same as debugging the server-side of any web application, so there is no need for special documentation. See the debugging instructions for your IDE or other tools. The following are examples and you can find much more on the web.

Reviewer: A couple suggestions for brevity. Feel free to take them or leave them.
```suggestion
Debugging the server-side code of an Office Add-in is the same as debugging the server-side of any web application. See the debugging instructions for your IDE or other tools. The following links will help you get started.
```

Author: > 

That makes it sound like all of these are relevant to everybody. I think the following will balance both our concerns:

```suggestion
Debugging the server-side code of an Office Add-in is the same as debugging the server-side of any web application. See the debugging instructions for your IDE or other tools. The following are examples for some of the most popular tools.
```

Reviewer: Yeah, that's a good solution.

- [Debug ASP.NET or ASP.NET Core apps in Visual Studio](/visualstudio/debugger/how-to-enable-debugging-for-aspnet-applications)
- [Debugging Express](https://expressjs.com/en/guide/debugging.html)
- [Node.js Debugging Guide](https://nodejs.org/en/docs/guides/debugging-getting-started/)
- [Node.js debugging in VS Code](https://code.visualstudio.com/docs/nodejs/nodejs-debugging)
- [Webpack Debugging](https://webpack.js.org/contribute/debugging/)

The rest of this article is concerned only with debugging client-side JavaScript (which may be transpiled from TypeScript).

> To find guidance for debugging client-side code, the first variable is the OS of your development computer.

Reviewer: Since this is the only time "OS" is used in this doc, consider spelling it out.
```suggestion
To find guidance for debugging client-side code, the first variable is the operating system of your development computer.
```

- [Windows](#debug-on-windows)
- [Mac](#debug-on-mac)
- [Linux or other Unix variant](#debug-on-linux)

## Debug on Windows

The following provides general guidance to debugging on Windows. There are special instructions for debugging UI-less custom functions in Excel and event-based add-ins in Outlook. See [Special cases in Windows](#special-cases-in-windows) later in this section.

> - If you're using Visual Studio, see [Debug Office Add-ins in Visual Studio](../develop/debug-office-add-ins-in-visual-studio.md).

Reviewer: Is there a way to restructure this list so it doesn't have to start with "If you're using"? It would be easier to scan if the bulleted list was like:
- **Visual Studio**:  ....
- **Visual Studio Code**: ...
- **Other IDEs**: ...

- If you're using Visual Studio Code, debug using the [Add-in Debugger Extension for Visual Studio Code](debug-with-vs-extension.md).

> - If you're using any other IDE or don't want to debug inside your IDE, use the developer tools that are associated with the browser runtime that add-ins use on your development computer. See one of the following:

Reviewer: Is there a case other than personal preference where someone would want to debug outside of their IDE?

Author: Yes. Their IDE might not support debugging client side code running in a dev server.

> [!TIP]
> [!INCLUDE[Identify the webview through the add-in UI](../includes/identify-webview-in-ui.md)]

### Special cases in Windows

To debug UI-less custom functions on Windows, see [UI-less custom functions debugging](../excel/custom-functions-debugging.md).

To debug event-based add-ins in Outlook, see [Debug your event-based Outlook add-in](../outlook/debug-autolaunch.md). The process requires Visual Studio Code.

## Debug on Mac

The following provides general guidance to debugging on Mac. There are special instructions for debugging UI-less custom functions in Excel. See [Special cases in Mac](#special-cases-in-mac) later in this section.

- If you're using Visual Studio Code, debug using the [Add-in Debugger Extension for Visual Studio Code](debug-with-vs-extension.md).
- For any other IDE, use the Safari Web Inspector. Instructions are in [Debug Office Add-ins on a Mac](debug-office-add-ins-on-ipad-and-mac.md).

### Special cases in Mac

To debug UI-less custom functions on Mac, see [UI-less custom functions debugging](../excel/custom-functions-debugging.md).

## Debug on Linux

> We don't recommend that you develop Office Add-ins on a Linux (or other Unix variant) computer except in the unusual case where you can be sure that all the add-ins users will be accessing the add-in through Office on the web from a Linux computer. You'll need to [sideload the add-in to Office on the web](sideload-office-add-ins-for-testing.md) to test and debug it. Debugging guidance is in [Debug add-ins in Office on the web](debug-add-ins-in-office-online.md).

Reviewer: Suggestion to make the opening point stand alone as a sentence. 
```suggestion
We don't recommend that you develop Office Add-ins on a Linux (or other Unix variant) computer. If you need to be sure that all the add-ins users can access the add-in through Office on the web from a Linux computer, you'll need to [sideload the add-in to Office on the web](sideload-office-add-ins-for-testing.md) to test and debug it. Debugging guidance is in [Debug add-ins in Office on the web](debug-add-ins-in-office-online.md).
```

Author: This would change the meaning. If their dev machine is Linux, the MUST sideload and debug in Office on the web. There's no other way because there's no desktop Office for Linux. The only situation in which we would recommend developing on Linux is when ALL of the end users of the add-in will themselves be using a Linux machine and, hence, using the add-in on Office on the web.

Reviewer: Ah, gotcha. I had misread that earlier. Consider highlighting that point of *all* the users being on Linux.
```suggestion
We don't recommend that you develop Office Add-ins on a Linux (or other Unix variant) computer. The only scenario that warrants developing add-ins on Linux is if all the add-in's users will access the add-in through Office on the web from a Linux computer. You'll need to [sideload the add-in to Office on the web](sideload-office-add-ins-for-testing.md) to test and debug it. Debugging guidance is in [Debug add-ins in Office on the web](debug-add-ins-in-office-online.md).
```

Author: My rewrite crossed your last suggestion "in the mail". I rewrote it a different way by putting the recommendation into a note at the end.
